dev/tools/controllerbuilder/pkg/llm/gemini.go

```diff
@@ -15,55 +15,215 @@
 package llm
 
 import (
-	"bytes"
 	"context"
 	"fmt"
 	"os"
-	"os/exec"
-	"strings"
 
-	"cloud.google.com/go/vertexai/genai"
-	"golang.org/x/oauth2/google"
+	"github.com/google/generative-ai-go/genai"
 	"google.golang.org/api/option"
 	"k8s.io/klog/v2"
 )
 
 // BuildGeminiClient builds a client for the Gemini API.
-func BuildGeminiClient(ctx context.Context) (*genai.Client, error) {
-	log := klog.FromContext(ctx)
-
+func BuildGeminiClient(ctx context.Context) (Client, error) {
 	var opts []option.ClientOption
 
 	if s := os.Getenv("GEMINI_API_KEY"); s != "" {
 		opts = append(opts, option.WithAPIKey(s))
-	} else {
-		// Some account can not use GEMINI_API_KEY but requires stricter access control via OAuth.
-		creds, err := google.FindDefaultCredentials(ctx, "https://www.googleapis.com/auth/generative-language", "https://www.googleapis.com/auth/cloud-platform")
+	}
+
+	client, err := genai.NewClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building gemini client: %w", err)
+	}
+
+	return &GeminiClient{
+		client: client,
+	}, nil
+}
+
+type GeminiClient struct {
+	client *genai.Client
+}
+
+func (c *GeminiClient) Close() error {
+	return c.client.Close()
+}
+
+func (c *GeminiClient) StartChat(systemPrompt string) Chat {
+	model := c.client.GenerativeModel("gemini-2.0-flash-exp")
+	// model := c.client.GenerativeModel("gemini-1.5-pro-002")
+
+	// Some values that are recommended by aistudio
+	model.SetTemperature(1)
+	model.SetTopK(40)
+	model.SetTopP(0.95)
+	model.SetMaxOutputTokens(8192)
+	model.ResponseMIMEType = "text/plain"
+
+	model.SystemInstruction = &genai.Content{
+		Parts: []genai.Part{
+			genai.Text(systemPrompt),
+		},
+	}
+
+	// if model.ToolConfig == nil {
+	// 	model.ToolConfig = &genai.ToolConfig{}
+	// }
+	// if model.ToolConfig.FunctionCallingConfig == nil {
+	// 	model.ToolConfig.FunctionCallingConfig = &genai.FunctionCallingConfig{}
+	// }
+	// model.ToolConfig.FunctionCallingConfig.Mode = genai.FunctionCallingAny
+
+	chat := model.StartChat()
+
+	return &GeminiChat{
+		model: model,
+		chat:  chat,
+	}
+}
+
+type GeminiChat struct {
+	model *genai.GenerativeModel
```

Comment:
But you cannot change the model after initiaizing the Chat. Wondering if we need to keep both attributes in this class?

Reply:
So it might not be the best API, but we can (and do) change the model... we call SetFunctionDefinitions